### PR TITLE
Update ChunkContainer.hpp

### PR DESCRIPTION
**Pull request by rghouzra**
# Description

This PR fixes the `ChunkContainer` for correct move semantics.

### CODE REVIEW

**Comment by xcelerit-team** [comment]
Thanks for submitting. The fixes for the move operations are clearly needed. 

The modifications of the `check_space` method are in partial conflict with the currently open PR #150, which is just about to get merged.

I suggest we leave this one open until #150 is merged, then rebase and finish it.

**Comment by auto-differentiation-dev** [comment]
Please try to update your branch after merging #150 .

**Comment by rghouzra** [comment]
> Please try to update your branch after merging #150 .

done

File: src/XAD/ChunkContainer.hpp [hunk 1]
```diff
@@ -133,24 +133,27 @@ class ChunkContainer
...
+        }
         return *this;
     }
-
     ChunkContainer(const ChunkContainer&) = delete;
     ChunkContainer& operator=(const ChunkContainer&) = delete;
 
     ~ChunkContainer() { _free_memory(); }
-
+
```
Comment (auto-differentiation-dev) [inline src/XAD/ChunkContainer.hpp:156]:
```suggestion

```

File: src/XAD/ChunkContainer.hpp [hunk 2]
```diff
@@ -436,23 +439,23 @@ class ChunkContainer
...
   private:
-    std::vector<char*> chunkList_;
-    size_type chunk_, idx_;
-
-    void check_space()
+  std::vector<char*> chunkList_;
+  size_type chunk_, idx_;
+  void check_space()
+{
+    if (XAD_VERY_LIKELY(chunk_ == chunkList_.size() - 1))
```
Comment (auto-differentiation-dev) [inline src/XAD/ChunkContainer.hpp:446]:
```suggestion
    std::vector<char*> chunkList_;
    size_type chunk_, idx_;
    void check_space()
    {
        if (XAD_VERY_LIKELY(chunk_ == chunkList_.size() - 1))
```

File: src/XAD/ChunkContainer.hpp [hunk 3]
```diff
@@ -438,21 +440,21 @@ class ChunkContainer
...
-            char* chunk = reinterpret_cast<char*>(
-                detail::aligned_alloc(ALIGNMENT, sizeof(value_type) * chunk_size));
-            if (chunk == NULL)
-                throw std::bad_alloc();
-            chunkList_.push_back(chunk);
+    {
+        char* chunk = reinterpret_cast<char*>(
+            detail::aligned_alloc(ALIGNMENT, sizeof(value_type) * chunk_size));
+        if (chunk == nullptr) {
+            throw std::bad_alloc();
```
Comment (auto-differentiation-dev) [inline src/XAD/ChunkContainer.hpp:450]:
```suggestion
        {
            char* chunk = reinterpret_cast<char*>(
                detail::aligned_alloc(ALIGNMENT, sizeof(value_type) * chunk_size));
            if (chunk == nullptr) {
                throw std::bad_alloc();
```

File: src/XAD/ChunkContainer.hpp [hunk 4]
```diff
@@ -438,21 +440,21 @@ class ChunkContainer
...
         }
-        ++chunk_;
-        idx_ = 0;
+        chunkList_.push_back(chunk);
```
Comment (auto-differentiation-dev) [inline src/XAD/ChunkContainer.hpp:452]:
```suggestion
            chunkList_.push_back(chunk);
```

File: src/XAD/ChunkContainer.hpp [hunk 5]
```diff
@@ -438,21 +440,21 @@ class ChunkContainer
...
     }
-
+    ++chunk_;
+    idx_ = 0;
+}
```
Comment (auto-differentiation-dev) [inline src/XAD/ChunkContainer.hpp:456]:
```suggestion
        ++chunk_;
        idx_ = 0;
    }
```

File: src/XAD/ChunkContainer.hpp [hunk 6]
```diff
@@ -438,21 +440,21 @@ class ChunkContainer
...
                 throw std::bad_alloc();
-            chunkList_.push_back(chunk);
         }
+            chunkList_.push_back(chunk);
+    }
         ++chunk_;
         idx_ = 0;
     }
-
+
```
Comment (auto-differentiation-dev) [inline src/XAD/ChunkContainer.hpp:457]:
```suggestion

```

File: src/XAD/ChunkContainer.hpp [hunk 7]
```diff
@@ -438,21 +440,21 @@ class ChunkContainer
...
     {
         if (XAD_VERY_LIKELY(chunk_ == chunkList_.size() - 1))
         {
             char* chunk = reinterpret_cast<char*>(
                 detail::aligned_alloc(ALIGNMENT, sizeof(value_type) * chunk_size));
-            if (chunk == NULL)
+            if (chunk == nullptr) {
                 throw std::bad_alloc();
-            chunkList_.push_back(chunk);
         }
```
Comment (auto-differentiation-dev) [inline src/XAD/ChunkContainer.hpp:451]:
```suggestion
            }
```

File: src/XAD/ChunkContainer.hpp [hunk 8]
```diff
@@ -438,21 +440,21 @@ class ChunkContainer
...
+            chunkList_.push_back(chunk);
+    }
```
Comment (auto-differentiation-dev) [inline src/XAD/ChunkContainer.hpp:453]:
```suggestion
        }
```

File: src/XAD/ChunkContainer.hpp [hunk 9]
```diff
@@ -438,21 +440,20 @@ class ChunkContainer
...
                 detail::aligned_alloc(ALIGNMENT, sizeof(value_type) * chunk_size));
-            if (chunk == NULL)
+            if (chunk == nullptr) {
                 throw std::bad_alloc();
+            }
             chunkList_.push_back(chunk);
         }
         ++chunk_;
         idx_ = 0;
     }
```
Comment (auto-differentiation-dev) [inline src/XAD/ChunkContainer.hpp:456]:
```suggestion
    }

```

File: src/XAD/ChunkContainer.hpp [hunk 10]
```diff
@@ -438,15 +440,15 @@ class ChunkContainer
   private:
     std::vector<char*> chunkList_;
     size_type chunk_, idx_;
-
     void check_space()
```
Comment (auto-differentiation-dev) [inline src/XAD/ChunkContainer.hpp:443]:
```suggestion

    void check_space()
```

File: src/XAD/ChunkContainer.hpp [hunk 11]
```diff
@@ -133,24 +133,26 @@ class ChunkContainer
...
+            chunkList_ = std::move(o.chunkList_);
+            chunk_ = o.chunk_;
+            idx_ = o.idx_;
+            o.chunk_ = 0;
+            o.idx_ = 0;
+        }
         return *this;
     }
-
     ChunkContainer(const ChunkContainer&) = delete;
```
Comment (auto-differentiation-dev) [inline src/XAD/ChunkContainer.hpp:152]:
```suggestion

    ChunkContainer(const ChunkContainer&) = delete;
```

File: src/XAD/ChunkContainer.hpp [hunk 12]
```diff
@@ -133,24 +133,26 @@ class ChunkContainer
...
     ChunkContainer& operator=(const ChunkContainer&) = delete;
 
     ~ChunkContainer() { _free_memory(); }
-
     void reserve(size_type s)
```
Comment (auto-differentiation-dev) [inline src/XAD/ChunkContainer.hpp:156]:
```suggestion

    void reserve(size_type s)
```

**Review by auto-differentiation-dev — APPROVED** [review]

**Pull request merged**